algorithms/linfa-nn/src/distance.rs

```diff
@@ -1,23 +1,21 @@
 use linfa::Float;
-use ndarray::Zip;
+use ndarray::{ArrayView, Dimension, Zip};
 use ndarray_stats::DeviationExt;
 
-use crate::Point;
-
 /// A distance function that can be used in spatial algorithms such as nearest neighbour.
-pub trait Distance<F: Float>: Clone {
+pub trait Distance<F: Float>: Clone + Send + Sync {
```

Comment:
it should also be possible to move the `D: Dimension` trait bound here to the trait definition and avoid declaring it twice , for each function once

Reply:
I need to use the same instance of `Distance` to compute distances on both `ArrayView1` and `ArrayView2`, so the methods themselves have to be generic with respect to the dimensionality.

Reply:
but the only flexibility you gain is that `distance` and `rdistance` can be called with different sized ndarrays, is that the intention?

Reply:
Yeah, I want the same instance of `Distance` to be able to call `distance(ArrayView1, ArrayView1)` and `distance(ArrayView2, ArrayView2)`. This is necessary in `KMeans`, where I use one instance of `Distance` to compute distances on 1D and 2D arrays. Note that `distance(ArrayView1, ArrayView2)` is not possible and the method panics if the lengths of the arrays don't match.